### DISCUSSION TITLE
Auto assignment with CODEOWNERS doesn't work

### DISCUSSION THREAD

**Discussion by derberg** [community/community#22523] [New to GitHub]
<p>Hi folks,</p>
<p>In the past when I had CODEOWNERS file in my repo and proper branch protection settings, when opening a PR, I could see people from CODEOWNERS automatically added to PR as reviewers.</p>
<p>Now problem is that I do not see this functionality to work. Am I doing something wrong, something changes?</p>
<p>Have a look for example at <a href="https://github.com/asyncapi/website/pull/512" class="inline-onebox" rel="noopener nofollow ugc">chore(blog): add release notes for 2.3 release by dalelane · Pull Request #512 · asyncapi/website · GitHub</a>.<br>
No reviewers are added even though we have <a href="https://github.com/asyncapi/website/blob/67be8f4b8aca32fb8b873cf6d47737012849aede/CODEOWNERS" class="inline-onebox" rel="noopener nofollow ugc">website/CODEOWNERS at 67be8f4b8aca32fb8b873cf6d47737012849aede · asyncapi/website · GitHub</a>. I would expect <code>alequetzalli</code> is added as reviewer by default. This user like other CODEOWNERS have “Maintain” rights.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/9/0/90fd60499f96d51a7da39b3ba331b1630287ce84.png" data-download-href="/uploads/short-url/kGDBUVVPVyqUugdU8xJ6RVaknt2.png?dl=1" title="Screenshot 2021-12-20 at 11.18.44" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/6995927/181071604-3eadcddc-d7e4-48ff-9dcf-404c080b1bc9.png" alt="Screenshot 2021-12-20 at 11.18.44" data-base62-sha1="kGDBUVVPVyqUugdU8xJ6RVaknt2" width="664" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/9/0/90fd60499f96d51a7da39b3ba331b1630287ce84_2_664x500.png, https://aws1.discourse-cdn.com/github/optimized/3X/9/0/90fd60499f96d51a7da39b3ba331b1630287ce84_2_996x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/9/0/90fd60499f96d51a7da39b3ba331b1630287ce84.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/9/0/90fd60499f96d51a7da39b3ba331b1630287ce84_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screenshot 2021-12-20 at 11.18.44</span><span class="informations">1311×987 126 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>Any ideas?</p>

**Comment by mszostok** [comment, marked as answer]
<p>Hi!</p>
<p>I will post my ideas here as it may help others. BTW it’s quite sad that no one from GH has responded since 21 Dec :slightly_frowning_face:  but let’s get straight to the point.</p>
<p>I did a short investigation and I spotted a few things:</p>
<ul>
<li>In all repositories, you added <code>@github-actions[bot]</code> in the <code>CODEOWNERS</code> file as an owner for a given pattern.</li>
<li>GitHub doesn’t highlight the <code>@github-actions[bot]</code>  owner, which is our first hint:<br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/d/5/d506012a351df04eecbec4116f0b0247fe540c47.png" data-download-href="/uploads/short-url/uouz5iH8sqO6viZyQmrtMb65hrx.png?dl=1" title="Screen Shot 2022-02-11 at 21.46.54" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/17568639/181071606-6c34fa6e-8ae7-4f73-875c-c6ba9522af65.png" alt="Screen Shot 2022-02-11 at 21.46.54" data-base62-sha1="uouz5iH8sqO6viZyQmrtMb65hrx" width="690" height="31" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/d/5/d506012a351df04eecbec4116f0b0247fe540c47_2_690x31.png, https://aws1.discourse-cdn.com/github/optimized/3X/d/5/d506012a351df04eecbec4116f0b0247fe540c47_2_1035x46.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/d/5/d506012a351df04eecbec4116f0b0247fe540c47.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/d/5/d506012a351df04eecbec4116f0b0247fe540c47_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screen Shot 2022-02-11 at 21.46.54</span><span class="informations">1071×49 11.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
</li>
<li>I checked whether they supported GitHub App Bots, and it looks like they don’t: <a href="https://github.community/t/is-it-possible-to-add-github-apps-bot-to-codeowners/3146" class="inline-onebox">Is it possible to add GitHub Apps Bot to CODEOWNERS?</a>
</li>
<li>Additionally, as you mentioned, each owner should be in a given organization (in this case, in <a href="https://github.com/asyncapi" rel="noopener nofollow ugc">AsyncAPI</a>), and have <code>write</code> permission. They can also be an outside collaborator :slight_smile:
</li>
</ul>
<p>Based on that, IMO the problem is the <code>@github-actions[bot]</code> in your <code>CODEOWNERS</code> file. In the official docs, you read:</p>
<blockquote>
<p>If any line in your CODEOWNERS file contains invalid syntax, the file will not be detected and will not be used to request reviews.</p>
</blockquote>
<p><strong>Suggested mitigation:</strong><br>
Get rid of <code>@github-actions[bot]</code> in your <code>CODEOWNERS</code> files. Additionally, get rid of <code>@github-actions[bot]</code>  for autoapprove. It’s just a matter of using a dedicated <a href="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token" rel="noopener nofollow ugc">PAT</a> for another account (a dedicated bot IMO) from your org instead of using <code>secrets.GITHUB_TOKEN</code>.</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/asyncapi/spec/blob/8ff3d0d30d23f152e520263ab7a8a4e305f52dc3/.github/workflows/automerge.yml#L21-L23">
  <header class="source">

      <a href="https://github.com/asyncapi/spec/blob/8ff3d0d30d23f152e520263ab7a8a4e305f52dc3/.github/workflows/automerge.yml#L21-L23" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/asyncapi/spec/blob/8ff3d0d30d23f152e520263ab7a8a4e305f52dc3/.github/workflows/automerge.yml#L21-L23" target="_blank" rel="noopener nofollow ugc">asyncapi/spec/blob/8ff3d0d30d23f152e520263ab7a8a4e305f52dc3/.github/workflows/automerge.yml#L21-L23</a></h4>



    <pre class="onebox">      <code class="lang-yml">
        <ol class="start lines" start="21" style="counter-reset: li-counter 20 ;">
            <li>uses: hmarr/auto-approve-action@v2</li>
            <li>with:</li>
            <li>  github-token: "${{ secrets.GITHUB_TOKEN }}"</li>
        </ol>
      </code>
    </pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>AFAIK, you cannot approve your own PRs (<a href="https://github.community/t/do-not-require-owner-approval-if-the-pull-request-is-from-an-owner/369/72" class="inline-onebox">Do not require owner approval if the pull request is from an owner? - #72 by borbudak-jadbio</a>). So, if you plan to use the already existing Bot <code>asyncapi-bot</code> for <a href="https://github.com/asyncapi/cli/pull/229" rel="noopener nofollow ugc">such PRs</a>, it probably won’t work :thinking:.</p>
<p><strong>UPDATE:</strong><br>
I’m even more confident about the proposed solution now, as I noticed it’s also described in the GitHub Action’s README that you use:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/hmarr/auto-approve-action#code-owners">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/hmarr/auto-approve-action#code-owners" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/17568639/181071607-148499dd-7bf5-429d-b82a-6da6a45009d9.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/1/c/1cde1ee4ed1462c3a2cb356addeb3ff23caf00fd_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/1/c/1cde1ee4ed1462c3a2cb356addeb3ff23caf00fd_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/1/c/1cde1ee4ed1462c3a2cb356addeb3ff23caf00fd.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/1/c/1cde1ee4ed1462c3a2cb356addeb3ff23caf00fd_2_10x10.png"></div>

<h3><a href="https://github.com/hmarr/auto-approve-action#code-owners" target="_blank" rel="noopener nofollow ugc">GitHub - hmarr/auto-approve-action: 👍 GitHub Action for automatically...</a></h3>

  <p>👍 GitHub Action for automatically approving GitHub pull requests - GitHub - hmarr/auto-approve-action: 👍 GitHub Action for automatically approving GitHub pull requests</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<blockquote>
<p>Rather than using a real user’s personal access token, you’re probably better off creating a dedicated bot user, and adding it to a team which you assign as the code owner. That way you can restrict the bot user’s permissions as much as possible, and your workflow won’t break when people leave the team.</p>
</blockquote>

**Comment by mszostok** [comment]
<p>BTW, I plan to add such features to old <a href="https://github.com/mszostok/codeowners-validator/" rel="noopener nofollow ugc"><code>codeowners-validator</code></a> tool:</p>
<ul>
<li>an option to format CODEOWNERS files.</li>
<li>an option to <strong>add/remove</strong> a given owner from CODEOWNERS files and create related PRs automatically. For example, with such (draft) config:<pre><code class="lang-yaml">updateRepositories:
  - path: "/tmp/cloned/repository" # local repository
  - url: "org/repo" # clone to temp dir
  - url: "org/repo2" # clone to temp dir

insert:
  - owners: [ "@hakuna", "@matata" ]
    condition:
      pattern: "*.go"
  - owners: [ "@baz", "@bar" ]
    condition:
      associatedWithOwner: "@ghost"

pullRequest:
  titleFormat: "Add {{ .NewOwner }} as a code owner in {{ .RepoName }}"
</code></pre>
The dedicated PRs will be created in your name if changes are detected.</li>
</ul>
<p>I see that the second option looks like sth that would be in your interest :thinking: as you have a lot of repos with CODEOWNERS files where you will need to perform such a fix.</p>
<p>I should be able to create some draft implementation soon if you are interested :smiley:.</p>
